File: internal/service/category.go
package service

import (
	"context"

	"things-expired/internal/model"
	"things-expired/internal/model/dto"
	"things-expired/internal/model/vo"
	"things-expired/internal/repository"
	"things-expired/pkg/errors"
	"things-expired/pkg/utils"
)

// ICategoryService 分类服务接口
type ICategoryService interface {
	Create(ctx context.Context, userID uint, req *dto.CreateCategoryRequest) (*vo.CategoryVO, error)
	List(ctx context.Context, userID uint, page, pageSize int) (*vo.CategoryListVO, error)
	Update(ctx context.Context, userID uint, req *dto.UpdateCategoryRequest) (*vo.CategoryVO, error)
	Delete(ctx context.Context, userID uint, categoryID uint) error
}

// CategoryService 分类服务实现
type CategoryService struct {
	categoryRepo repository.ICategoryRepository
}

// NewCategoryService 创建分类服务
func NewCategoryService(categoryRepo repository.ICategoryRepository) ICategoryService {
	return &CategoryService{categoryRepo: categoryRepo}
}

func (s *CategoryService) Create(ctx context.Context, userID uint, req *dto.CreateCategoryRequest) (*vo.CategoryVO, error) {
	category := &model.Category{
		UserID:    userID,
		Name:      req.Name,
		Color:     req.Color,
		Icon:      req.Icon,
		SortOrder: req.SortOrder,
	}

	if category.Color == "" {
		category.Color = "#000000"
	}

	if err := s.categoryRepo.Create(ctx, category); err != nil {
		return nil, err
	}

	return s.toVO(category), nil
}

func (s *CategoryService) List(ctx context.Context, userID uint, page, pageSize int) (*vo.CategoryListVO, error) {
	if page <= 0 {
		page = 1
	}
	if pageSize <= 0 {
		pageSize = 10
	}

	categories, total, err := s.categoryRepo.List(ctx, userID, page, pageSize)
	if err != nil {
		return nil, err
	}

	list := make([]vo.CategoryVO, 0, len(categories))
	for _, c := range categories {
		list = append(list, *s.toVO(c))
	}

	return &vo.CategoryListVO{
		List:  list,
		Total: total,
	}, nil
}

func (s *CategoryService) Update(ctx context.Context, userID uint, req *dto.UpdateCategoryRequest) (*vo.CategoryVO, error) {
	category, err := s.categoryRepo.GetByID(ctx, req.CategoryID)
	if err != nil {
		return nil, err
	}
	if category == nil {
		return nil, errors.New(errors.CodeParamInvalid, "分类不存在")
	}

	// 检查权限
	if category.UserID != userID {
		return nil, errors.New(errors.CodeForbidden, "无权限操作此分类")
	}

	if req.Name != "" {
		category.Name = req.Name
	}
	if req.Color != "" {
		category.Color = req.Color
	}
	if req.Icon != "" {
		category.Icon = req.Icon
	}
	category.SortOrder = req.SortOrder

	if err := s.categoryRepo.Update(ctx, category); err != nil {
		return nil, err
	}

	return s.toVO(category), nil
}

func (s *CategoryService) Delete(ctx context.Context, userID uint, categoryID uint) error {
	category, err := s.categoryRepo.GetByID(ctx, categoryID)
	if err != nil {
		return err
	}
	if category == nil {
		return errors.New(errors.CodeParamInvalid, "分类不存在")
	}

	// 检查权限
	if category.UserID != userID {
		return errors.New(errors.CodeForbidden, "无权限操作此分类")
	}

	return s.categoryRepo.Delete(ctx, categoryID)
}

func (s *CategoryService) toVO(category *model.Category) *vo.CategoryVO {
	return &vo.CategoryVO{
		CategoryID: category.ID,
		UserID:     category.UserID,
		Name:       category.Name,
		Color:      category.Color,
		Icon:       category.Icon,
		SortOrder:  category.SortOrder,
		CreatedAt:  utils.FormatTimeUTC(category.CreatedAt),
	}
}

File: src/service/category.ts
import type { ICategoryRepository, IItemRepository } from "@/repository/interfaces";
import type { ICategoryService } from "@/service/interfaces";
import type { CategoryVO } from "@/model/vo/category";
import type {
  CreateCategoryRequest,
  UpdateCategoryRequest,
} from "@/model/dto/category";
import { toCategoryVO } from "@/model/vo/category";
import { AppError } from "@/errors";
import {
  CodeCategoryNotFound,
  CodeDatabaseError,
  CodeForbidden,
  CodeUserExists,
} from "@/errors/code";
import { nowUTC } from "@/utils/time";

export class CategoryService implements ICategoryService {
  constructor(
    private categoryRepo: ICategoryRepository,
    private itemRepo: IItemRepository,
  ) {}

  async create(
    userId: number,
    req: CreateCategoryRequest,
  ): Promise<CategoryVO> {
    // 检查同名分类
    const existing = await this.categoryRepo.findByUserIdAndName(
      userId,
      req.name,
    );
    if (existing) {
      throw new AppError(CodeUserExists, "分类名称已存在");
    }

    const now = nowUTC();
    const category = await this.categoryRepo.create({
      userId,
      name: req.name,
      color: req.color || null,
      icon: req.icon || null,
      sortOrder: req.sort_order ?? 0,
      createdAt: now,
      updatedAt: now,
    });

    return toCategoryVO(category);
  }

  async list(
    userId: number,
    params: { page: number; pageSize: number; keyword?: string },
  ): Promise<{ list: CategoryVO[]; total: number; page: number }> {
    const { items, total } = await this.categoryRepo.list({
      userId,
      ...params,
    });

    return {
      list: items.map(toCategoryVO),
      total,
      page: params.page,
    };
  }

  async update(
    userId: number,
    req: UpdateCategoryRequest,
  ): Promise<CategoryVO> {
    const category = await this.categoryRepo.findById(req.category_id);
    if (!category) {
      throw new AppError(CodeCategoryNotFound, "分类不存在");
    }
    if (category.userId !== userId) {
      throw new AppError(CodeForbidden, "无权操作此分类");
    }

    // 检查分类名是否重复
    if (req.name && req.name !== category.name) {
      const existing = await this.categoryRepo.findByUserIdAndName(
        userId,
        req.name,
      );
      if (existing && existing.id !== req.category_id) {
        throw new AppError(CodeUserExists, "分类名称已存在");
      }
    }

    const updateData: Record<string, any> = {};
    if (req.name !== undefined) updateData.name = req.name;
    if (req.color !== undefined) updateData.color = req.color;
    if (req.icon !== undefined) updateData.icon = req.icon;
    if (req.sort_order !== undefined) updateData.sortOrder = req.sort_order;
    updateData.updatedAt = nowUTC();

    const updated = await this.categoryRepo.update(req.category_id, updateData);
    if (!updated) {
      throw new AppError(CodeDatabaseError, "更新分类失败");
    }

    return toCategoryVO(updated);
  }

  async delete(userId: number, categoryId: number): Promise<void> {
    const category = await this.categoryRepo.findById(categoryId);
    if (!category) {
      throw new AppError(CodeCategoryNotFound, "分类不存在");
    }
    if (category.userId !== userId) {
      throw new AppError(CodeForbidden, "无权操作此分类");
    }

    // 先删除分类下的所有物品
    await this.itemRepo.deleteByCategoryId(categoryId);
    // 再删除分类
    await this.categoryRepo.delete(categoryId);
  }
}
